File: app/src/main/java/za/org/grassroot2/services/MediaServiceImpl.java
package za.org.grassroot2.services;

import android.content.ContentResolver;
import android.content.Context;
import android.content.Intent;
import android.database.Cursor;
import android.media.MediaScannerConnection;
import android.net.Uri;
import android.os.Environment;
import android.provider.MediaStore;
import android.support.v4.content.FileProvider;
import android.text.TextUtils;
import android.webkit.MimeTypeMap;
import android.widget.ImageView;

import java.io.File;
import java.io.IOException;
import java.text.SimpleDateFormat;
import java.util.Date;

import javax.inject.Inject;

import io.reactivex.Single;
import io.reactivex.schedulers.Schedulers;
import timber.log.Timber;
import za.org.grassroot2.dagger.ApplicationContext;
import za.org.grassroot2.database.DatabaseService;
import za.org.grassroot2.model.MediaFile;
import za.org.grassroot2.model.exception.FailedToCreateMediaFileException;
import za.org.grassroot2.util.ImageUtil;

/**
 * Created by luke on 2017/08/11.
 */

public class MediaServiceImpl implements MediaService {

    private static final SimpleDateFormat IMG_TIMESTAMP = new SimpleDateFormat("yyyyMMdd_HHmmss");
    public static final int DESIRED_COMPRESSED_WIDTH = 1280;
    public static final int DESIRED_COMPRESSED_HEIGHT = 720;
    public static final String FILEPROVIDER_AUTHORITY = "za.org.grassroot2.fileprovider";

    private final Context applicationContext;
    private final DatabaseService databaseService;
    private final NetworkService networkService;
    private final ImageUtil imageUtil;

    @Inject
    public MediaServiceImpl(@ApplicationContext Context applicationContext,
                            DatabaseService databaseService,
                            NetworkService networkService, ImageUtil imageUtil) {
        this.applicationContext = applicationContext;
        this.databaseService = databaseService;
        this.networkService = networkService;
        this.imageUtil = imageUtil;
    }

    @Override
    public Single<String> createFileForMedia(final String mimeType, final String mediaFunction) {
        return Single.create(e -> {
            try {
                File imageFile = createImageFile(mimeType);
                final Uri imageUri = FileProvider.getUriForFile(applicationContext,
                        FILEPROVIDER_AUTHORITY,
                        imageFile);
                Timber.d("taking image, URI = " + imageUri);
                MediaFile toSave = new MediaFile(imageUri.toString(), imageFile.getAbsolutePath(), mimeType, mediaFunction);
                toSave.setCompressOnSend(true);
                MediaFile createdFile = databaseService.storeObject(MediaFile.class, toSave);
                Timber.d("created media file = " + createdFile);
                final String createdUid = createdFile.getUid();
                e.onSuccess(createdUid);
            } catch (Throwable t) {
                Timber.e(t);
                throw new FailedToCreateMediaFileException();
            }
        });
    }

    private File createImageFile(final String mimeType) throws IOException {
        final String timeStamp = IMG_TIMESTAMP.format(new Date());
        final String extension = "." + (TextUtils.isEmpty(mimeType) ? "jpg" : MimeTypeMap.getSingleton().getExtensionFromMimeType(mimeType));
        File storageDirectory = applicationContext.getExternalFilesDir(Environment.DIRECTORY_PICTURES);
        if (!storageDirectory.exists() && !storageDirectory.mkdirs()) {
            Timber.e("Could not create picture folder");
        }
        return File.createTempFile(
                "IMG_" + timeStamp + "_",
                extension,
                storageDirectory
        );
    }

    @Override
    public Single<String> captureMediaFile(final String mediaFileUid) {
        return Single.create(e -> {
            MediaFile mediaFile = databaseService.loadObjectByUid(MediaFile.class, mediaFileUid);
            mediaFile.setReadyToUpload(true);
            if (mediaFile.getMimeType().startsWith("image") && mediaFile.isCompressOnSend()) {
                imageUtil.resizeImageFile(mediaFile.getAbsolutePath(), mediaFile.getAbsolutePath(), DESIRED_COMPRESSED_WIDTH, DESIRED_COMPRESSED_HEIGHT);
            }
            e.onSuccess("DONE");
            MediaScannerConnection.scanFile(applicationContext, new String[] { mediaFile.getAbsolutePath() }, new String[] {mediaFile.getMimeType()}, (path, uri) -> Timber.d(path, uri.toString()));
        });
    }

    @Override
    public Single<String> storeGalleryFile(final String mediaFileUid, final Uri fileUri) {
        return Single.create(e -> {
            final MediaFile mediaFile = databaseService.loadObjectByUid(MediaFile.class, mediaFileUid);
            String localFileName = getLocalFileNameFromURI(fileUri);
            if (mediaFile.isCompressOnSend()) {
                imageUtil.resizeImageFile(localFileName, mediaFile.getAbsolutePath(), DESIRED_COMPRESSED_WIDTH, DESIRED_COMPRESSED_HEIGHT);
            } else {
                mediaFile.setAbsolutePath(localFileName);
                mediaFile.setContentProviderPath(fileUri.toString());
            }
            mediaFile.setMimeType(getMimeType(fileUri));
            mediaFile.setReadyToUpload(true);
            e.onSuccess("DONE");
            databaseService.storeObject(MediaFile.class, mediaFile);
        });
    }

    private String getLocalFileNameFromURI(final Uri selectedImage) {
        String localImagePath = null;
        String[] filePathColumn = {MediaStore.Images.Media.DATA};
        Cursor cursor = applicationContext.getContentResolver().query(selectedImage, filePathColumn, null, null, null);
        if (cursor != null) {
            cursor.moveToFirst(); // if null, will throw error to subscriber, so check in here would be redundant
            int columnIndex = cursor.getColumnIndex(filePathColumn[0]);
            localImagePath = cursor.getString(columnIndex);
            cursor.close();
        }
        return localImagePath;
    }

    private String getMimeType(Uri uri) {
        String mimeType = null;
        String extension;
        if (uri.getScheme().equals(ContentResolver.SCHEME_CONTENT)) {
            final MimeTypeMap mime = MimeTypeMap.getSingleton();
            extension = mime.getExtensionFromMimeType(applicationContext.getContentResolver().getType(uri));
        } else {
            extension = MimeTypeMap.getFileExtensionFromUrl(Uri.fromFile(new File(uri.getPath())).toString());
        }
        switch (extension) {
            case "jpg":
            case "jpeg":
                mimeType = "image/jpeg";
                break;
            case "png":
                mimeType = "image/png";
                break;
            default:
                break;
        }

        return mimeType;
    }
}

File: app/src/main/java/za/org/grassroot2/services/MediaServiceImpl.kt
package za.org.grassroot2.services

import android.content.ContentResolver
import android.content.Context
import android.media.MediaScannerConnection
import android.net.Uri
import android.os.Environment
import android.support.v4.content.FileProvider
import android.text.TextUtils
import android.webkit.MimeTypeMap
import io.reactivex.Single
import timber.log.Timber
import za.org.grassroot2.dagger.ApplicationContext
import za.org.grassroot2.database.DatabaseService
import za.org.grassroot2.model.MediaFile
import za.org.grassroot2.model.exception.FailedToCreateMediaFileException
import za.org.grassroot2.util.FileUtil
import za.org.grassroot2.util.ImageUtil
import java.io.File
import java.io.IOException
import java.text.SimpleDateFormat
import java.util.*
import javax.inject.Inject

/**
 * Created by luke on 2017/08/11.
 */

class MediaServiceImpl @Inject
constructor(@param:ApplicationContext private val applicationContext: Context,
            private val databaseService: DatabaseService,
            private val imageUtil: ImageUtil) : MediaService {


    override fun storeMediaFile(fileName: String): Single<String> {
        return Single.create { e ->
            try {
                val mediaFile = MediaFile(fileName, fileName, "", MediaFile.FUNCTION_LIVEWIRE)
                val stored = databaseService.storeObject(MediaFile::class.javaObjectType, mediaFile)
                e.onSuccess(stored.uid)
            } catch (t: Exception) {
                t.printStackTrace()
                e.onError(t)
            }
        }
    }

    override fun createFileForMedia(mimeType: String, mediaFunction: String): Single<String> {
        return Single.create { e ->
            try {
                val imageFile = createImageFile(mimeType)
                val imageUri = FileProvider.getUriForFile(applicationContext,
                        FileUtil.FILEPROVIDER_AUTHORITY,
                        imageFile)
                Timber.d("taking image, URI = " + imageUri)
                val toSave = MediaFile(imageUri.toString(), imageFile.absolutePath, mimeType, mediaFunction)
                toSave.isCompressOnSend = true
                val createdFile = databaseService.storeObject(MediaFile::class.java, toSave)
                Timber.d("created media file = " + createdFile)
                val createdUid = createdFile.uid
                e.onSuccess(createdUid)
            } catch (t: Throwable) {
                Timber.e(t)
                throw FailedToCreateMediaFileException()
            }
        }
    }

    @Throws(IOException::class)
    private fun createImageFile(mimeType: String): File {
        val timeStamp = IMG_TIMESTAMP.format(Date())
        val extension = "." + if (TextUtils.isEmpty(mimeType)) "jpg" else MimeTypeMap.getSingleton().getExtensionFromMimeType(mimeType)
        val storageDirectory = applicationContext.getExternalFilesDir(Environment.DIRECTORY_PICTURES)
        if (!storageDirectory!!.exists() && !storageDirectory.mkdirs()) {
            Timber.e("Could not create picture folder")
        }
        return File.createTempFile(
                "IMG_" + timeStamp + "_",
                extension,
                storageDirectory
        )
    }

    override fun captureMediaFile(mediaFileUid: String, targetImgWidth: Int, targetImgHeight: Int): Single<String> {
        return Single.create { e ->
            val mediaFile = databaseService.loadObjectByUid(MediaFile::class.java, mediaFileUid)
            mediaFile!!.setReadyToUpload(true)
            if (mediaFile.mimeType.startsWith("image") && mediaFile.isCompressOnSend) {
                imageUtil.resizeImageFile(mediaFile.absolutePath, mediaFile.absolutePath, targetImgWidth, targetImgHeight)
            }
            e.onSuccess("DONE")
            MediaScannerConnection.scanFile(applicationContext, arrayOf(mediaFile.absolutePath), arrayOf(mediaFile.mimeType)) { path, uri -> Timber.d(path, uri.toString()) }
        }
    }

    override fun captureMediaFile(mediaFileUid: String): Single<String> {
        return this.captureMediaFile(mediaFileUid, DESIRED_COMPRESSED_WIDTH, DESIRED_COMPRESSED_HEIGHT)
    }

    override fun storeGalleryFile(mediaFileUid: String, fileUri: Uri, targetImgWidth: Int, targetImgHeight: Int): Single<String> {
        return Single.create { e ->
            val mediaFile = databaseService.loadObjectByUid(MediaFile::class.java, mediaFileUid)
            val localFileName = FileUtil.getLocalFileNameFromURI(fileUri, applicationContext)
            localFileName?.let {
                if (mediaFile!!.isCompressOnSend) {
                    imageUtil.resizeImageFile(it, mediaFile.absolutePath, targetImgWidth, targetImgHeight)
                } else {
                    mediaFile.absolutePath = localFileName
                    mediaFile.contentProviderPath = fileUri.toString()
                }
                mediaFile.mimeType = getMimeType(fileUri)
                mediaFile.setReadyToUpload(true)
                e.onSuccess("DONE")
                databaseService.storeObject(MediaFile::class.java, mediaFile)
            }
        }
    }

    override fun storeGalleryFile(mediaFileUid: String, fileUri: Uri): Single<String> {
        return this.storeGalleryFile(mediaFileUid, fileUri, DESIRED_COMPRESSED_WIDTH, DESIRED_COMPRESSED_HEIGHT)
    }

    private fun getMimeType(uri: Uri): String? {
        var mimeType: String? = null
        val extension: String
        if (uri.scheme == ContentResolver.SCHEME_CONTENT) {
            val mime = MimeTypeMap.getSingleton()
            extension = mime.getExtensionFromMimeType(applicationContext.contentResolver.getType(uri))
        } else {
            extension = MimeTypeMap.getFileExtensionFromUrl(Uri.fromFile(File(uri.path)).toString())
        }
        when (extension) {
            "jpg", "jpeg" -> mimeType = "image/jpeg"
            "png" -> mimeType = "image/png"
            else -> {
            }
        }

        return mimeType
    }

    companion object {

        private val IMG_TIMESTAMP = SimpleDateFormat("yyyyMMdd_HHmmss")
        val DESIRED_COMPRESSED_WIDTH = 1280
        val DESIRED_COMPRESSED_HEIGHT = 720
    }
}